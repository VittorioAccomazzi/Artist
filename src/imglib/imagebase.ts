import { textChangeRangeIsUnchanged } from "typescript";


interface TypedArray {
    readonly length : number;
    [n: number]: number;
    buffer : ArrayBuffer;
}

interface TypedArrayConstructor<T extends TypedArray> {
    new (buffer: ArrayBuffer, byteOffset?: number, length?: number): T;
}

export type ImageUint8  = ImageBase<Uint8Array,Uint8ArrayConstructor>
export type ImageUint16 = ImageBase<Uint16Array,Uint16ArrayConstructor>
export type ImageFloat32= ImageBase<Float32Array,Float32ArrayConstructor>

export class ImageFactory {

    static Uint8( width : number, height : number ) : ImageUint8 {
        return new ImageBase<Uint8Array,Uint8ArrayConstructor>(Uint8Array.BYTES_PER_ELEMENT, Uint8Array, width, height)
    }
    static Uint16( width : number, height : number ) : ImageUint16{
        return (new ImageBase<Uint16Array,Uint16ArrayConstructor>(Uint16Array.BYTES_PER_ELEMENT, Uint16Array, width, height)) 
    }
    static Float32( width : number, height : number ): ImageFloat32{
        return new ImageBase<Float32Array,Float32ArrayConstructor>(Float32Array.BYTES_PER_ELEMENT, Float32Array, width, height)
    }
}


class ImageBase<T extends TypedArray, C extends TypedArrayConstructor<T>> {

    private buffer : T
    private pixels : T [] 

    constructor( elementSize : number, constructor : C, width : number, height : number ) {
        let nPixels= width * height;
        let bytes  = nPixels * elementSize
        let buffer= new ArrayBuffer(bytes);
        this.pixels= Array.from(Array(height), (e,i)=> new constructor (buffer, width*elementSize*i, width) ) 
        this.buffer= new constructor(buffer, 0, nPixels)
    }


    /**
     * get type of pixel stored.
     */
    get imageType(): string {
        return this.pixels[0].constructor.name.replace('Array','')
    }

    /**
     * returns the current list of pixels.
     */
    get imagePixels(): T {
        return this.buffer 
    }

    /**
     *  get the image buffer
     */
    get imageBuffer() : ArrayBuffer {
        return this.buffer.buffer
    }

    /**
     * get image height
     */
    get height() : number {
        return this.pixels.length
    }

    /**
     * return image height
     */
    get width() : number {
        return this.pixels[0].length
    }

    /**
     * get a pixel value.
     * @param x row of the pixel queried
     * @param y column of the pixel queried
     */
    get( x: number, y: number ) : number {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates queried: image size ${this.width}x${this.height} pixel ${x},${y}`)
        return this.pixels[y][x]
    }

    /**
     * get a row of pixels
     * @param y row of pixels to select.
     */
    getRow(y:number) : T {
        if( y<0 || y>this.height )throw new Error (`Invalid y coordinate queried : image height ${this.height} y ${y}`)
        return this.pixels[y]
    }

    /**
     * 
     * @param x row pixel to set
     * @param y column pixel to se
     * @param value  value to set
     */
    set(x:number, y:number, value : number) : void {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates set: image size ${this.width}x${this.height} pixel ${x},${y}`)
        this.pixels[y][x]= value
    }

    /**
     * check that the value passed are valid image coordinates.
     * @param x row
     * @param y column
     */
    private checkBoundary(x: number, y : number ) : boolean {
        return x>=0 && x < this.width && y>=0 && y<this.height
    }

}