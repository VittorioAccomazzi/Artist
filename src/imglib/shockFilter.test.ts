import { getCanvases, hash, dumpCanvas, toSeqCanvas, dumpImage, toNodeCanvas } from './testutils'
import CanvasUtils from './canvasUtils'
import { ImageFactory, ImageFloat32 } from './imagebase'
import ShockFilter from './shockFilter'

test('shall enhance edges', async ()=>{
    let width = 200
    let height= 280
    let bck = 50
    let frg = 100
    let image = ImageFactory.Uint8(width,height)
    let pixels= image.imagePixels
    pixels.forEach((v,i)=>{
        let y = (i/width | 0) 
        let x = (i%width)
        let val = bck
        if( Math.abs(x-y)<8 ){
            val=frg+(x | 0)
        }
        pixels[i]=val
    })

    // await dumpImage(image, 'Shock input image')
    let shkImage = ShockFilter.Run(image, 2, 0.0001, 2 )
    // await dumpImage(logImage, 'Sock  image')
    let hsh = await hash(shkImage)
    expect(hsh).toMatchSnapshot();

})


test('shall enhance edges',async()=>{
    for await ( const [canvas,ctx,name] of getCanvases() ) {
        let [lImg, aImg, bImg ] = CanvasUtils.toLab(toSeqCanvas(canvas))
        
        let shkImg = ShockFilter.Run(lImg, 2, 0.0001, 4) as ImageFloat32
        // await dumpImage( shkImg, `Shock image ${name}`)
        let hsh = await hash(shkImg)
        expect(hsh).toMatchSnapshot()

        let sCanvas = CanvasUtils.fromLab(shkImg, aImg, bImg)
        // dumpCanvas(toNodeCanvas(sCanvas), `Shock enhannced ${name}`)
        hsh = await hash(toNodeCanvas(sCanvas))
        expect(hsh).toMatchSnapshot()
    }
},60*2*1000)