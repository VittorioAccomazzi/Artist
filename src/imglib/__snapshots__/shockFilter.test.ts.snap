// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shall enhance edges 1`] = `"fe000000000000000000000000000000ff000000000000000000000000000000ff800000000000000000000000000000ffe00000000000000000000000000000fff000000000000000000000000000007ffc00000000000000000000000000001ffe000000000000000000000000000007ff000000000000000000000000000003ffc00000000000000000000000000001ffe00000000000000000000000000000fff800000000000000000000000000003ffc00000000000000000000000000000fff000000000000000000000000000007ff800000000000000000000000000003ffc00000000000000000000000000000ffe000000000000000000000000000007ff800000000000000000000000000001ffe00000000000000000000000000000fff000000000000000000000000000007ff800000000000000000000000000001ffc00000000000000000000000000000fff000000000000000000000000000003ffc00000000000000000000000000001ffe00000000000000000000000000000fff000000000000000000000000000003ff800000000000000000000000000001ffe000000000000000000000000000007ff800000000000000000000000000003ffc00000000000000000000000000001ffe000000000000000000000000000007ff000000000000000000000000000003ffc00000000000000000000000000000fff000000000000000000000000000007ff800000000000000000000000000003ffc00000000000000000000000000000fff000000000000000000000000000007ff800000000000000000000000000003ffe00000000000000000000000000000fff000000000000000000000000000007ff800000000000000000000000000001ffe00000000000000000000000000000fff000000000000000000000000000007ffc00000000000000000000000000001ffe00000000000000000000000000000fff000000000000000000000000000003ffc00000000000000000000000000001ffe00000000000000000000000000000fff000000000000000000000000000003ffc00000000000000000000000000000ffe000000000000000000000000000007ff800000000000000000000000000003ffc00000000000000000000000000001ffe000000000000000000000000000007ff800000000000000000000000000001ffc00000000000000000000000000000fff000000000000000000000000000007ff800000000000000000000000000003ffc00000000000000000000000000000fff000000000000000000000000000003ff800000000000000000000000000001ffe00000000000000000000000000000fff000000000000000000000000000007ff800000000000000000000000000001ffe000000000000000000000000000007ff000000000000000000000000000003ffc00000000000000000000000000001ffe00000000000000000000000000000fff000000000000000000000000000003ff000000000000000000000000000001ff0000000000000000000000000000007f0000000000000000000000000000003f0000000000000000000000000000000f0000000000000000000000000000000f0000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall enhance edges 2`] = `"000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000084000000000000000000000000000007ffff40000000000000000000000001fffffff800000000000000000000000fffffffdfe800000000000000000000fffffffffff800000000000000000007fffffffffffc0000000000000000007ffffffffffffff00000000000000002ffffffffffffffff000000000000000ffffffffffffffffff00000000000005fffffffffffffffffff000000000003ffffffffffffffffffffe0800000000ffffffffffffffffffffff3000000003ffffffffffffffffffffffe00000003ffffffffffffffffffffffff4000002bffffffffffffffffffffffffff00003fffffffffffffffffffffffffffd200ffffffffffffffffffffffffffffff00ffffffffffffffffffffffffffffffa0ffffffffffffffffffffffffffffffe0fffffffffffffffffffffffffffffff0fffffffffffffffffffffffffffffff8fffffffffffffffffffffffffffffffcdffffffffffffffffffffffffffffffeffffffffffffffffffffffffffffffff00000000000000000000000000000000000000000000000000000040000000000000000000002bfcc000004000000000000000000001fffffc0000000000000000000000001fffffff1f00000000000000000000007fffffffcc1800000000000000000003fffffffffe1c0000000000000000001ffffffffffff0000000000000000000ffffffffffff800000000000000000007fffffffffff80000000000000000001ffffffffffff80000000000000000003ffffffffffffe000000000000000000ffffffffffffffc00000000000000003ffffffffffffffe0000000000000001ffffffffffffffffc000000000000003fffffffffffffffffc0000000000000fffffffffffffffffff0000000000001fffffffffffffffffffe000000000007ffffffffffffffffffff40000000003fffffffffffffffffffffe0000000003ffffffffffffffffffffffe00000000ffffffffffffffffffffffff00000001ffffffffffffffffffffffffe0000003fffffffffffffffffffffffff000001ffffffffffffffffffffffffffc00007ffffffffffffffffffffffffffffd07ffffffffffffffffffff7fffd000000fffffffffffffffffffe620130000000fffffffffffffffffff0000000000000ffffffffffffffffff80000000000000fffffffffffdfc000000000000000000fffffffffff8d0000000000000000000fffffffffffff3ffdfa000000000e000fffffffffffd806074f000018e000000fffffffffffc005c004c000000000000fffffffff8007fa00000000000000000fffe0000000000000000000000000000000000000000000000000000000000000000000000000003fffffffc0000000000000000000002ffff7fffffffffffff0000000012effffffffffffffffffffb000000003ffffffffffffffffffffffb00000007ffffffffffffffffffffff5f00000007fffffffffffffffffc53ffff0000004ffffffffffffffffffffffffb10161ffffffffffffffffffffefe3ff47017fffffffffffffffffffffc00800083fffffffffffffffffffdf800000000ffffffffffffffffff7ffffd00000000bffffffffffffffffffffe0000000000ffffffffffffffffffffc20000000000fffffffffffffffffffffff380000000ffffffffffffffffffffe00000000000fffffff7e7fd17ffffffff4000000000ffefffffff1ffffe03ffe000000000007fcbffffa43ff81ffffc000000000000dfffff001fe7ffff0000304000000000e7eb7078683ffd700030000000000000001bffc1efff80077c020000000000000000c37ffe000efff000000000000000800c0e780004f04000000000000000000003ff001fffffff007fc10000000000001a0002fff100000e0000000000000003801d07f800000f81f0000000000000ffffffffffffffffffffffffffffffeffffffffffffffffff4bffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdfffffffffffff8fffffffffffffc7ffffffffffffffbeffffffffffffffffff7fffff07ffffffffffffffffffffffffffffffffbffffffffffffffffc0441c7efffff1fffffffdfffffffff0ffffffffffff7c3fffffeffffff801ff9e3ffe077ffffffffffffffdfffffffaf020700087fefffffffd37ffffffefffffffffde7f0ff1e7e1dff9ffffffffdf000003fffffeeff91ffffffff37ffe00004380fffe101ffffffa03301e00001fffff0c000075f9fe7fffffffffff80c7ff01f3c000003ff02001f8e3f0041c0000007fff00030007c7fffffffe2100740308000010000000080000fff03cfff5fc03ff80000000d0000000ffffc10000000000400000018ff8070f000004000700f800000018807ffffff0000000000001800000004200117ff8074800000063fc0000300000fd02004000a38000ff9f82000100000004000000001ef00003e00000000000000000000000ff80000000000000000000000000000080000000000000000000000020000000000f8700000000000000000000000000fff8406300000000000000000000000f000100180000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall enhance edges 3`] = `"0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000147d9000000000000000000000000005fffffff000000000000000000000004fffffffdf80000000000000000000003fffffffffd200000000000000000007fffffffffffc0000000000000000007fffffffffffff900000000000000003ffffffffffffffff0000000000000007fffffffffffffffff80000000000005ffffffffffffffffffc004000000003ffffffffffffffffffffc0800000000fffffffffffffffffffffff800000000fffffffffffffffffffffff40000007fffffffffffffffffffffffffc00000fffffffffffffffffffffffffff6600bfffffffffffffffffffffffffffd700ffffffffffffffffffffffffffffff90ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffe0fffffffffffffffffffffffffffffff2ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000020000000000000000000000000000003e0000000000000000000003b7b000001c000000000000000000001fffffa0000000000000000000000001ffffffe1f00000000000000000000007fffffffcc1800000000000000000003ffffffffff7c0000000000000000000ffffffffffff00000000000000000007fffffffffff80000000000000000001ffffffffffff80000000000000000001ffffffffffffe0000000000000000007fffffffffffff000000000000000001ffffffffffffffc0000000000000000fffffffffffffffe0000000000000001ffffffffffffffffc000000000000003fffffffffffffffff80000000000001fffffffffffffffffff8000000000003ffffffffffffffffffff000000000007ffffffffffffffffffff00000000001fffffffffffffffffffffe0000000003ffffffffffffffffffffff800000000ffffffffffffffffffffffff00000003ffffffffffffffffffffffff80000007ffffffffffffffffffffffffc000001ffffffffffffffffffffffffffc0000fffffffffffffffffffffffffffded07ffffffffffffffffffff7fffd000000fffffffffffffffffffe620130000000fffffffffffffffffff0000000000000ffffffffffffffffff80000000000000fffffffffffdfc000000000000000000fffffffffff8d0000000000000000000fffffffffffff3ffdfa000000000a000fffffffffffc00607470000186000000fffffffffff4005c004c000000000000fffffffff8007c000000000000000000fffe0000000000000000000000000000000000000000000000000000000000000000000000000003fffffffc0004000000000000000002ffff7fffffffffffff0000000012efffffffffffffffffffff000000003fffffffffffffffffffffff00000017ffffffffffffffffffffff7f00000007fffffffffffffffffc53ffff0000006ffffffffffffffffffffffffb101f1ffffffffffffffffffffffe3ffe7817fffffffffffffffffffffc00800183fffffffffffffffffffdf800000000ffffffffffffffffff7ffffd00000000fffffffffffffffffffffe0000000000ffffffffffffffffffffc20000000000fffffffffffffffffffffff380000000ffffffffffffffffffffe00000000000fffffff7e7fd07ffffffff4000000000ffefffffff1ffffe03ffa00000000000ffcbffffa43ff80ffffc000000000000dfffff001fe6ffff0000300000000000e7fb7078683ffd700010000000000000101bffc1efff80077c020000000000000400c3fffe000afff000000000000000c08c0e780004f04000000000000000000003ff801ffffff6007fc10000000000001a0002fff100000a0000000000000003801d8ff800000f80f0000000000000ffffffffffffffffffffffffffffffffffffffffffffffffd43ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdfffffffffffff9fffffffffffffc7ffffffffffffffbfffffffffffffffffff5fffff07ffffffffffffffffffffffffffffffffbffffffffffffffffc0441c7cfffff1fffffffdfffffffff0ffffffffffff7c1fffffeffffff801ff9e3ffe07fffffffffffffffdfffffffef024700087fefffffffd37ffffffefffffffffde7f0ff1e7e1dff9ffffffffdf000003fffffeeff10ffffffff33ffe00004b80fffe101ffffffa03301e01061fffff0c000075f9fc7fffffffffff80c7ff01f3c000003ff02001f8e3f0041c0000007fff00030003c7fffffffe2100740308000010000000080000fff03cfffffc03ff8100000050000000ffffc10000000400e80000018ff8070f000004000700f800000018807ffffff000000000000180000000400011fff0074000040063fc0000300000fd00000000a38000ff9f82000100000004000000001ef0001be00000000000000000000000e780000000000000000000000000000080000000000000000000000020000000000f8700000000000000000000000000fff8406300000000000000000000000f000100180000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall enhance edges 4`] = `"fffe0000b5ffffffbc007fffffc00003fffe0000bfffffff90003fffffc00000fffe00003ffffedfe0003fffffe00018fffe00001fffffdfe0001fffffe00030fffe00001fffffdfe0d01ffffff000e0fffe00002ffffbcfe8001ffffff001c0fffe000015ffc007c0001ffffff80180fffe000003f9c00bc0001ffffff80300fffe000001b7c002c0000ffffffc0600fffe0000000c0000c8000ffffffe0400fffe00000002100004000ffffffe0800fffe000001388ff800000fffffff0000fffe00000018ffff02000fffffff0000fffe00000901ffffc0000fffffff8000fdfe00000079ffffe0000fffffff8000f9fe000001e13ffff8000fffffff8000f1fe000001017ffffc000fffffff0000f1fe000002016ffffe000ffffffe0000e1fe00000600ffffff800fe7fffc0000e1fe000004007fffffc00fe3fffc0001c1fe000000005fffffe00fe3fff8000381fe000000007ffffff00fe1fff0000781fe00000000fffffff00fe0ffe0000f01fe00000000fffffff80fe0ffe0001f01fe00000001ffdffffe0fe07fc0001f01fe00004003ffffffff0fe07fc0003f01fe0000c007f8ffffff0fe03f80007f01fe00008006f1ffffff8fe01f0000ff01fe0001800fc7ffffff8fe03f0001ff01fe0001801f8fffffffcfe07f0001ff01fe000180031fffffffcfe1ff0003ff01fe000180043fffffffe7e3ff0007ff00fe000380007ffffffff7c7ff0007ff00fe00030000ffffffffff9fff000fff00fe00030001ffffffffffffff000fff00fe00030007dfffffffffffff001fff00fe000380073fffffffffffff003fff00fe0003800f7ffffffffffffe003fff00fe0003c01ceffffffffffffe007fff00fe0003e029fffffffffffff6007fff00fe0007e043ffffffffffffec00ffff01fe0007e08737fffffffffffc01ffff01fe0007f00a2e61ffffffffcc01ffff01fe0007f808580c7e7fffff1803ffff01fe0007f810120e007ffffc1803ffff01fe0003c000240000fffff03007ffff01fe0003d0004000017fffc0e007ffff01fe0003d800800002ffff03c007ffff01fe0003f800400005fffc3f800fffff01fe0001f00000001ffff03f000fffff00fe0001f00000000fffe03e0007ffff00fe0000f00000000ffff0380007ffff00fe0000f00000001ffff0300007ffff00fe0000f00000003ffff8300007ffff00fe0000e00000007ffff830000fffff00fe000060000005bffff830000fffff00fe00106000000bbffffc30005fffff00fe000060000007fffffc10001fffff00fe00002000001fbffffc1000bfffff00fe00000000001ffffffc1000bfffff00fe0000b000001feffffc10003fffff00fe00018800003f81fffc10007ffffc00fe00011800007f787fe010007ffffc00fe0003800001fe003fc018007ffffc00fe000fc00003fc013f801802ffffff00fe003f800003f808bf001803ffffff00fe001f000007f00c3e180807fffff800fe000e000003e01c1e180807ffffe300fe003c00400fc67e3e780807ffffdf00ff00d00040079f871f80080fffff7f00ff0380c0401f1fff7fe0080bfffeff00ff000001002e1ffffff0080ffffdff00ff038000801c1fffeff0080fffffff00ff0e0000005c3ffffff80c1fffffff00ff00000010781ffffffc0c1ffff7ff00ff01000008b03fffdff80c1fffefff00ff0c000028f03fffcff80c3fffefff00ff10010030601fffcff80c3fffdfff00ff04010080c01ffe4ff80c3fffdfff00ff20010043c01ffef7f0067fffbfff00ff00100007801ffe87f0067fffbfff00ff00100087001fff03f0067fffbfff01ff0000c41b001fffffe0067fffffff01ff04004076001fffffe0067fff7fff01ff080060e4001fffffc004ffff7fff00ff10003068001fffffc006ffff7fff01ff200011c8001ffffec002ffff7fff01ff080005900017c0008003ffff7fff01ff480405a00007e0ff8003ffff7fff01ff10040f600003f3ff0003fffe7fff01ff20080fc00001f8fe0003fffe7fff00ff220823800001ff1e0003fffe7fff00ff200903000000fffc0003fffe7fff00ff000913000000fffc0003fffe7fff00ff0008880000007ff80003fffc7fff01ff0008280000001ff80003fffcfffe01ff700c0f0000001ff80003fffefff001ffd00c2d800003fffe00031ffeffe000ff800664c00007ffff8003c07effc000ff200241800007ffffe003f800ff80c0ff400115a00003fffff803fe04ff00e0ffc1400e100007fffffc07ff84fe00f0ff800107500007fffffe07ffedfc0078ffc00003a80007ffffff07fffffc003cff800003980043ffffff07fffff8003cff800000880047ffffff87fffff8003cff800001980007ffffff87fffff8003cff800001c80047ffffffc7fffff0003cff8040019800c7ffffffc7fffff0003cff000009f001c7ffffffe7fffff0003cff8000073c01c7ffffffe7fffff0001cff0000038181c7ffffffe7fffff0001cff0000004001cfffffffe7ffffe0001cff80000730018ffffffff7ffffc0001cff80000130038ffffffff8ffffc0001cff80000800039ffffffff81fffc0001cff00000840071ffffffff801ff80001eff00000c00063ffffffff8007f80001e7f800006000e7ffffffff8007f00001e7f80000340187ffffffffc007e00001e7e8000008010fffffffffc007c00001e7e8000000001fffffffffc007800001e7f0000010003fffffffffc007000000e7f000201800ffffffffffe004000000e7e000001001ffffffffffe020000800f7e000104007ffffffffffe070801800f7f000000007fffffffffff07b803000f7e00000000ffffffffffff07f80300"`;

exports[`shall enhance edges 5`] = `"fffe0000a5fffdffbd007fffffc00003fffe0000adfffcdf90003fffffc00000fffe00000bfffccfc0003fffffc00018fffe000019ffffdfe1001fffffe00030fffe00001ffffe5fe0c01ffffff000e0fffe00000bfdfbcfa0001ffffff001c0fffe0000051bc007c0001ffffff80180fffe000003f9c003c0001ffffff80300fffe000001bdc003c0001ffffffc0600fffe0000000c8001c0001ffffffe0c00fffe00000004800184000ffffffe0800fffe000001bc8ff804000fffffff0000fffe00000298ffff0e000fffffff0000fffe00000903ffffc0000fffffff8000f9fe0000007bffffe0000fffffff8000f9fe000001f1fffff8000fffffff8000f1fe000001037ffffc000fffffff0000f1fe00000203fffffe000ffffffe0000e1fe00000602ffffff800ff7fffc0000e1fe000004017fffffc00fe3fffc0001c1fe000000007fffffe00fe3fff8000381fe00000000fffffff00fe1fff0000781fe00000000fffffff00fe0ffe0000f01fe00000001fffffff80fe0ffe0001f01fe00000003fffffffe0fe07fc0003f01fe0000400fffffffff0fe07f80003f01fe0000c00ff8ffffff0fe03f80007f01fe0000800ff3ffffff8fe01f0000ff01fe0001801fc7ffffff8fe03f0001ff01fe0001801f8fffffffcfe0ff0003ff01fe000180231fffffffcfe1ff0003ff00fe000180443fffffffe7e3ff0007ff00fc000180007ffffffff7cfff0007ff00fc00030000fffffffffb9fff000fff00fe00030001ffffffffffffff001fff00fe00030007dfffffffffffff001fff00fe000380063fffffffffffff003fff00fe0003800f7ffffffffffffe003fff00fe0003801ceffffffffffffe007fff00fe0003c029fffffffffffff6007fff00fe0007c043ffffffffffffec00ffff00fe0007c08737fffffffffffc01ffff00fe0007f00a2e71ffffffffcc01ffff00fe0007f008580cfe7fffff1803ffff00fe0007d800120e007ffffc1803ffff00fe0003c000244000ffffe03007ffff00fe0003c0004800017fffc0e007ffff00fe0003d800900802ffff03c007ffff00fe0003e00040100dfffc7f800fffff00fe0001f00000001ffff13f000fffff00fe0001f00000000fffe03c000fffff00fe0000f00000005ffff0380007ffff00fe0000f00000003ffff0700007ffff00fe0000f00000003ffff8300007ffff00fe0000e0000000fffff830000fffff00fe000060000085fffff830004fffff00fe00106000000bfffffc30005fffff00fe000060000017fffffc10001fffff00fe00002100001fbffffc1000bfffff00fe00000400001ffffffc1000bfffff00fe0000b000001feffffc10003fffff00fe00010800003f81fffc10017ffffe00fe00011000007f787fe018007ffffc00fe0003d00001fe003fc018007ffffc00fe001fc00003fc013f801803ffffff00fe003f804003f808bf001803ffffff00fe001f000007f00c3e180807ffffff00fe000c000003e01c1e180807fffff300fe003c00400fc67e1e780807ffffdf00fe00f100c0079f871f80080fffffff00fe03c0c0401f1fff7fe0080bfffeff00fe010001002e1ffffff0080ffffdff00fe07c000c01c1fffeff0080fffffff00fe0f0180005c1fffeff80c1fffffff00fe00020010381fffeff80c1ffff7ff00fe01000008b03fffcff80c1fffefff00fe2c410028701fffcff80c3fffefff00fe1c810031601fff8ff80c3fffdfff00fe04090080c01ffe4ff80c3fffdfff00fe20010043c00ffe77f0067fffbfff00ff00914007800ffe87f0067fffbfff00ff0010c087000fff03f0067fffbfff00ff0000c41b000fffffe0067fffffff01ff0c404076000fffffe006ffff7fff01ff088060e4000fffffc004ffff7fff01ff10003068000fffffc006ffff7fff01ff200011c80007effec006ffff7fff01fe080407900007c0008003ffff7fff01ff480405a00003e0ff8003ffff7fff01ff10040f600001f1ff0003fffe7fff00ff22086fc00001f87e0003fffe7fff00ff220923800000fc1e0003fffe7fff00ff200903000000fffc0003fffe7fff00ff00091b000000fffc0003fffe7fff00ff0008d80000007ff80003fffe7fff00ff0008780000001ff80003fffcfffe01ff700c9f0000001ff80003fffefff001ffd00e2d800003fffe00031ffeffe000ff800664c00007ffff8003c07effc000ff2002c5a00007ffffe003f800ff80c0ff4023d7a00003fffff803fe04ff00e0ffc1408f500003fffffc07ff84fe00f0ff800107500003fffffe07ffe5fc0078ffc00107a80007ffffff07fffdfc003cff800007980043ffffff07fffff8003cff800001980047ffffff87fffff8003cff800011980007ffffff87fffff8003cff804001d80007ffffffc7fffff0003cff804001980047ffffffc7fffff0003cff008009f000c7ffffffe7ffdfe0003cff8000073c00c7ffffffe7fffff0001cff8000038181c7ffffffe7ffffc0001cff000003c001cfffffffe7ffffc0001cff80000730018ffffffff7ffffc0001cff80000130018ffffffff8ffffc0001cff80000810011ffffffff81fffc0001c7f80000840011ffffffff801ff80001e7f00000c80063ffffffff8007f00001e7e80000600067ffffffff8007f00001e7e80000340087ffffffffc007e00001e7e8000008010fffffffffc007c00003e7e8000020000fffffffffc007000001e7e0002110003fffffffffc006000000e7f00024b8007fffffffffe004000400e7e000121000ffffffffffe020000800f7e000124000ffffffffffe060001800f7f000080003fffffffffff071803000f7e000000007fffffffffff07f00300"`;
